### ISSUE TITLE
Partial override of default configuration not available when using TypeScript

### ISSUE THREAD

**Issue by nullromo** [winstonjs/winston-daily-rotate-file#305]
Because the typing for `DailyRotateFileTransportOptions` includes `RequireOnlyOne<GeneralDailyRotateFileTransportOptions, 'filename' | 'stream'>`, TypeScript will complain if neither a `filename` nor a `stream` is specified in the options object.

In the README, `filename` is listed as having a default value of `'winston.log.%DATE%` (which I'm pretty sure is a typo [no closing single-quote] anyway), and `stream` is listed as default `null`. From reading the README, I expect that creating a transport like this:

```ts
winston.createLogger({
    transports: [
        new DailyRotateFile({
            dirname: 'logs',
        }),
    ],
});
```

will work, allowing me to use the default `filename` option with my own custom `dirname` option. However, this does not work. TypeScript says that the type `{ dirname: string; }` is not assignable to the parameter. Fixing it like this:

```ts
new DailyRotateFile({
    filename: 'winston.log.%DATE%',
    dirname: 'logs',
}),
```

works, of course, but in the absence of a custom `filename`, the default should be allowed, even with other custom options like `dirname`.

**Comment by mattberther** [comment]
@nullromo Would you please take a look at #307 and let me know if that will solve this issue for you. Also, as Im not familiar with TypeScript, I'll see if @Slessi is also available for a quick sanity check on the approach in that PR before I merge.

**Comment by nullromo** [comment]
@mattberther That idea seems to work. I tried out the code and it worked for my purposes, however I did not extensively test it. I also created another pull request #308 with a minor edit. Thank you for looking into this!

**Comment by Slessi** [comment]
This would be a regression.

See relevant conversation https://github.com/winstonjs/winston-daily-rotate-file/pull/238#discussion_r324716271
And TS playground test [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8mwEsD2A7AzlAvFA4hFEATgIYA2ciqGAPrPMugOoLAAWAYgqfsQLbS0KDNMzYBlYIQh8A3ACg5CFMCIAzYgGNoQqqI5ce-KBAAeKlABMMeAiXL0qUAN5yobqKoMo+EAFxQ0SSUAc3l3AMlpXgB+fwIANyJ5AF8FJRVCdS06SiYWVgkpPmMzfCtcfCIyHXRnV3dPbm9+WKgEpPq3QKLef26QlLTlNU1oGyr7XIwXcNUkJD6glFC5VLlVAFcUDVyoFUCACiQHdH8atABKOvDjqYA6bqiw91vhO8bDCEG5feADpw8838AHJgVBkhd5L9-oCFlBQQAaDxeHwgsEQqEQQ4AuZwxERHpo8GQn5Yv44oHw4FIj7NPxUpGPPhEjEKOQAenZUCkqiI+C0Fg8hCQvCgrGAwDAaF8nMCmgA1khEplSEgAO53DQi9kARw2ZOE7IALAAGACsAEYTQA2C0W9mgSBoDSEBDwAC06RGWndUj1CCk7tQECDqndwDVSHdYGFkEIiCx4ajpgQgR+4GgACUIP6pDAUKQQPmIAAeAAqSIA0hAQBhTOZyvKa0hVFAy1goE2QC22wA+LCdKAABQQGnl5aRAFETBpSBsLKWuz2K1Bq7Xe73BwAya7hdwAbUrUCUq5raAAuu7YoO9+5s7mIBYSyOxxPVxub7eoDuh8R4wgyBLbMtUIJ9p1nedSzXNAq17JEtgXTwCAsD9wmSQ8z3PeQFEdbQTjQGBXWCJQyA7e8NgDCB80LYsS3GOxzkIhBiO8UgkWBWkfDBWhgSZXhgV7bCvUyUYKlsap8KYljSJmBoURaRZXWWZ4ukiPhWn6ZTB1xRSBlWBRNm2XZfikkjSCOfCzkkoizKuWS3FeKgHjU3gVKgRz0HeeSvn00lAlM1iYR0qliUxfybMCik8WpZEmlRELWRMiKyCCyl8T4lkSSS5izNS6KaW8tFGRczLsKAA)

Original situation achieves its intended goal, one of stream/filename is required (I realise this disagrees with your situation)

![image](https://user-images.githubusercontent.com/3223500/109956826-668f7f00-7ce4-11eb-988d-198c3b1cc401.png)

With this change, the situation is worse however. It is allowed to pass neither, but now also allowed to pass both, and the function isnt automatically aware of either parameter.

![image](https://user-images.githubusercontent.com/3223500/109956902-7d35d600-7ce4-11eb-9532-16b2f406cfa8.png)

**Comment by nullromo** [comment]
Ah yes, this change does mean `const a: Options = {filename: '', stream: ''}` stops giving an error, which is not acceptable. I am fairly new to TypeScript myself. Perhaps [this question](https://stackoverflow.com/questions/66480691/does-typescript-have-a-way-to-require-property-a-nand-property-b) will yield a solution.

**Comment by Slessi** [comment]
@nullromo probably all that is required here is modifying `RequireOnlyOne` by wrapping the following:

```diff
{
    [K in Keys]-?:
        Required<Pick<T, K>>
        & Partial<Record<Exclude<Keys, K>, undefined>>
}[Keys]
```
with `Partial<>`
```diff
Partial<{
    [K in Keys]-?:
        Required<Pick<T, K>>
        & Partial<Record<Exclude<Keys, K>, undefined>>
}[Keys]>;
```

**Comment by nullromo** [comment]
I submitted #309, which provides a working solution and is a little less verbose.

**Comment by Slessi** [comment]
@nullromo do as I suggested above ^ it still doesnt work

![image](https://user-images.githubusercontent.com/3223500/110028339-68346380-7d33-11eb-9e7e-c1ba0265e64b.png)

**Comment by nullromo** [comment]
I botched it with a typo again. [This](https://www.typescriptlang.org/play?#code/C4TwDgpgBA6glsAFgewK7ADwBUA0UCqAfFALxQDeUA2gApRwB2UAogB4DGANqgCYQYBrCCGQAzKLihCR4ogF0A-AC4oDCADcIAJygBfANwAoUJCgANAPIAlbHiKkoACixQAPgQCUUCK2AQGPADOUMgARgBWEOzAUApO8EhomJL2AGSebvEIKOgY+HhYxOlYXiou7vhGhox+WqIAhuzQAOL+2vWcFmDAcMgMweSGUMNQPHBaDPUAthAqgcBajADmRrrVDLUNTVBdPX2BCYgAYnCc-tPQPn4Bwa1qWh27vf0UQyOip+czylDziwwrQxrGraLbQJ77Q4AZQWEGm3l8-iCUDu7U63WeAzewz+cKmPz+y1WhkMAHpSSZoABZEBYcDQMiWGwQ-qHE5nSYzPAsg7ZGFaPGEfRQckScDIFQ8ZAQQIMADkMQA7sgtAJjPSdhj9g4mRhUQ90Xt+nhdTy2Z9ORBuVrWXzYdNCEKRaSoAAJeqaKDAZBQVCBaBMr2KuBNEnsfYxeoqHkOShjCYXFRyuV6YUjdMZ4aiwI5Tg8KChcEAaUM4f6MVC0ZtwTIcfGlqTcrwHw5iagydTzt+ufzhZ2JbL8yg7CrRprFFG9bbybwuOmjc7WZdObQeYLxdLEdGo8xscnCZmjebFunTd+9qmC4MXZXqDXffqTG0WhVhiAA) is what I meant to do (see lines 14 and 15). I updated the PR.

**Comment by Slessi** [comment]
@nullromo My personal opinion is that you should still just remove `Required` from the `Pick`.

**Comment by nullromo** [comment]
The PR is updated.

**Comment by Slessi** [comment]
kinda confusing which PR to check but #309 seems ok

**Comment by mattberther** [comment]
Resolved with #307 and pushed to npm as `winston-daily-rotate-file@4.5.1`.
